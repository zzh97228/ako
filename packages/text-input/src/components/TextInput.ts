import vue, { computed, defineComponent, h, mergeProps, reactive, ref, toRef, VNodeArrayChildren } from 'vue';
import { useModel, genModelProps, useFieldConsumer, genColorProp, useColor } from '@lagabu/shared';

export default defineComponent({
  name: 'TextInput',
  props: {
    small: Boolean,
    large: Boolean,
    flat: Boolean,
    placeholder: String,
    disabled: Boolean,
    prefixDivider: Boolean,
    suffixDivider: Boolean,
    ...genModelProps([String, Number]),
    ...genColorProp('primary'),
  },
  setup(props, context) {
    const { emit } = context;
    const { class: colorClasses, style: colorStyles } = useColor(props, true);
    const { model, lazyState, setInnerState } = useModel(props, context);
    const { onBlur, onFocus, state: fieldState, hasError } = useFieldConsumer({ model, lazyState, setInnerState });
    const inputRef = ref<null | HTMLInputElement>(null);
    const state = reactive({
      isComposing: false,
    });
    const onInput = (e: InputEvent) => {
      if (state.isComposing) return;
      const target = e.target as HTMLInputElement;
      model.value = target.value;
      return target.value;
    };

    const genInput = () =>
      h('input', {
        class: 'text-input__content',
        type: 'text',
        ref: inputRef,
        disabled: props.disabled,
        value: lazyState.value,
        placeholder: props.placeholder,
        onCompositionstart: () => {
          state.isComposing = true;
        },
        onCompositionend: () => {
          state.isComposing = false;
        },
        onFocus: onFocus,
        onBlur: onBlur,
        onChange: (e: InputEvent) => {
          if (props.disabled) {
            e.preventDefault();
            e.stopImmediatePropagation();
            return;
          }
        },
        onInput: (e: InputEvent) => {
          if (props.disabled) {
            e.preventDefault();
            e.stopImmediatePropagation();
            return;
          }
          onInput(e);
          emit('input', e, onInput);
        },
      });

    function createAddonSlot(className: string, prefix: boolean) {
      return (children?: VNodeArrayChildren) =>
        h(
          'div',
          {
            class: {
              [className]: true,
              'text-input--divider': prefix ? props.prefixDivider : props.suffixDivider,
            },
          },
          children
        );
    }

    return {
      genInput,
      genInputPrefix: createAddonSlot('text-input__prefix', true),
      genInputSuffix: createAddonSlot('text-input__suffix', false),
      inputRef,
      isFocusing: toRef(fieldState, 'isFocusing'),
      colorClasses,
      colorStyles,
      hasError,
    };
  },
  render() {
    return h(
      'div',
      mergeProps(
        {
          class: {
            'text-input__wrapper': true,
            'text-input--flat': this.flat,
            'text-input--small': this.small,
            'text-input--large': this.large,
            'text-input--focusing': this.isFocusing,
            'text-input--error': this.hasError,
          },
        },
        {
          class: this.colorClasses,
          style: this.colorStyles,
        }
      ),
      [
        this.$slots.prefix && this.genInputPrefix(this.$slots.prefix()),
        this.genInput(),
        this.$slots.suffix && this.genInputSuffix(this.$slots.suffix()),
      ]
    );
  },
});
