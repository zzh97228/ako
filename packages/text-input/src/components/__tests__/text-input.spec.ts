import { mount, VueWrapper } from '@vue/test-utils';
import { defineComponent } from 'vue';
import TextInput from '../TextInput';
describe('input.ts', () => {
  let mountFunc: (options?: object) => VueWrapper<any>;
  beforeEach(() => {
    mountFunc = (opts = {}) => mount(TextInput, opts);
  });
  it('should render input component', () => {
    const wrapper = mountFunc({
      props: {
        modelValue: 'hello world',
      },
    });
    expect(wrapper.classes()).toContain('text-input__wrapper');
    expect(wrapper.find('input').classes()).toContain('text-input__content');
    const inputElm = wrapper.find('input').element;
    expect(inputElm.getAttribute('type')).toBe('text');
    expect(inputElm.value).toBe('hello world');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should change its value when change `v-model` value', async () => {
    const wrapper = mount(
      defineComponent({
        components: {
          TempInput: TextInput,
        },
        data() {
          return {
            txt: 'hello',
          };
        },
        template: `<temp-input v-model="txt"></temp-input>`,
      })
    );
    await wrapper.setData({
      txt: 'world',
    });
    expect(wrapper.find('input').element.value).toBe('world');
  });
});
