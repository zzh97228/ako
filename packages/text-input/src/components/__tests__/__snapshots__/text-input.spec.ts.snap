// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input.ts should render input component 1`] = `
<div class="text-input__wrapper">
  <input class="text-input__content"
         type="text"
  >
</div>
`;
